Networking problems after second kubeone run

I am using kubeone to set up a cluster on Hetzner cloud. After the initial run, things mostly work. There is a problem reaching kube-dns-upstream from pods on the worker nodes, but that can be fixed by doing a rolling restart of coredns.

However, after I run kubeone for a second time, canal gets redeployed, which then renders the cluster unusable. Every request seems to be taking 10 seconds (which of course makes debugging a pain).

This is how I execute kubeone (on both runs). `kubeone.yaml ` and `tf.json` are identical on both runs. I am using kubeone 1.2.2 (and have also tried 1.2.1) and have tried with kubernetes 1.20 as well.
```
kubeone appl --auto-approve --debug --manifest kubeone.yaml -t tf.json
```

```
-- kubeone.yaml
apiVersion: kubeone.io/v1beta1
kind: KubeOneCluster
versions:
    kubernetes: 1.19.11
cloudProvider:
  hetzner: {}
  external: true
containerRuntime:
  containerd: {}
```


```
-- tf.json
{
  "kubeone_api": {
    "sensitive": false,
    "type": [
      "object",
      {
        "endpoint": "string"
      }
    ],
    "value": {
      "endpoint": "x.x.x.x"
    }
  },
  "kubeone_hosts": {
    "sensitive": false,
    "type": [
      "object",
      {
        "control_plane": [
          "object",
          {
            "bastion": "string",
            "cloud_provider": "string",
            "cluster_name": "string",
            "network_id": "string",
            "private_address": [
              "tuple",
              [
                "string",
                "string",
                "string"
              ]
            ],
            "public_address": "dynamic",
            "ssh_agent_socket": "string",
            "ssh_port": "number",
            "ssh_private_key_file": "string",
            "ssh_user": "string"
          }
        ]
      }
    ],
    "value": {
      "control_plane": {
        "bastion": "xxx.xxx.xxx.xxx",
        "cloud_provider": "hetzner",
        "cluster_name": "test-001",
        "network_id": "xxx",
        "private_address": [
          "x.x.x.x",
          "x.x.x.x",
          "x.x.x.x"
        ],
        "public_address": null,
        "ssh_agent_socket": "env:SSH_AUTH_SOCK",
        "ssh_port": 22,
        "ssh_private_key_file": "/builds/infrastructure/infrastructure.tmp/TF_VAR_CLUSTER_PRIVATE_KEY",
        "ssh_user": "root"
      }
    }
  },
  "kubeone_workers": {
    "sensitive": false,
    "type": [
      "object",
      {
        "test-001-pool1": [
          "object",
          {
            "providerSpec": [
              "object",
              {
                "cloudProviderSpec": [
                  "object",
                  {
                    "firewall": "string",
                    "image": "string",
                    "labels": [
                      "object",
                      {
                        "test-001-workers": "string"
                      }
                    ],
                    "location": "string",
                    "networks": [
                      "tuple",
                      [
                        "string"
                      ]
                    ],
                    "serverType": "string"
                  }
                ],
                "operatingSystem": "string",
                "operatingSystemSpec": [
                  "object",
                  {
                    "distUpgradeOnBoot": "bool"
                  }
                ],
                "sshPublicKeys": [
                  "tuple",
                  [
                    "string"
                  ]
                ]
              }
            ],
            "replicas": "number"
          }
        ]
      }
    ],
    "value": {
      "test-001-pool1": {
        "providerSpec": {
          "cloudProviderSpec": {
            "image": "ubuntu-20.04",
            "labels": {
              "test-001-workers": "pool1"
            },
            "location": "nbg1",
            "networks": [
              "x"
            ],
            "serverType": "ccx12"
          },
          "operatingSystem": "ubuntu",
          "operatingSystemSpec": {
            "distUpgradeOnBoot": false
          },
          "sshPublicKeys": [
            "ssh-ed25519 xxxxxxx"
          ]
        },
        "replicas": 1
      }
    }
  }
}
```

Does anyone have an idea what might be causing this? What logs should I look at specifically for debugging?

@kron4eg Thank you for the swift reply and PR. We did build kubeone ourselves and have since migrated to the released v1.2.3. However, we still see the same thing happening.

1. We deploy the cluster.
2. We start a container on a worker node, log in and can do `nslookup google.com` just fine, we get fast responses
3. We run kubeone a second time. All the canal pods get terminated and restarted.
4. We try to do `nslookup` again and get no response `;; connection timed out; no servers could be reached`
5. Looking at the node-local-dns logs on the worker node, we see that the kube-dns-upstream pod cannot be reached: `[ERROR] plugin/errors: 2 route53.amazonaws.com.cluster.local. A: dial tcp 10.106.52.115:53: i/o timeout`

This only happens after the second kubeone run.

--- a/pkg/state/context.go
+++ b/pkg/state/context.go
@@ -90,7 +90,6 @@ type State struct {
 	ForceUpgrade              bool
 	ForceInstall              bool
 	UpgradeMachineDeployments bool
-	PatchCNI                  bool
 	CredentialsFilePath       string
 	ManifestFilePath          string
 	PauseImage                string

--- a/pkg/tasks/cni.go
+++ b/pkg/tasks/cni.go
@@ -51,12 +51,3 @@ func ensureCNIExternal(s *state.State) error {
 	s.Logger.Infoln("External CNI plugin will be used")
 	return nil
 }
-
-func patchCNI(s *state.State) error {
-	if !s.PatchCNI {
-		return nil
-	}
-
-	s.Logger.Info("Patching CNI...")
-	return ensureCNI(s)
-}

--- a/pkg/tasks/tasks.go
+++ b/pkg/tasks/tasks.go
@@ -217,10 +217,6 @@ func WithResources(t Tasks) Tasks {
 				Predicate:   func(s *state.State) bool { return s.Cluster.CloudProvider.External },
 			},
 			{
-				Fn:     patchCNI,
-				ErrMsg: "failed to patch CNI",
-			},
-			{
 				Fn:     joinStaticWorkerNodes,
 				ErrMsg: "failed to join worker nodes to the cluster",
 			},

--- a/pkg/templates/canal/canal.go
+++ b/pkg/templates/canal/canal.go
@@ -150,7 +150,7 @@ func Deploy(s *state.State) error {
 		configMap(buf, s.Cluster.ClusterNetwork.CNI.Canal.MTU),
 		daemonsetServiceAccount(),
 		deploymentServiceAccount(),
-		daemonSet(s.PatchCNI, s.Cluster.ClusterNetwork.PodSubnet, calicoCNIImage, calicoNodeImage, flannelImage),
+		daemonSet(s.Cluster.ClusterNetwork.PodSubnet, calicoCNIImage, calicoNodeImage, flannelImage),
 		controllerDeployment(calicoControllerImage),
 	)
 

--- a/pkg/templates/canal/daemonset.go
+++ b/pkg/templates/canal/daemonset.go
@@ -26,7 +26,7 @@ import (
 
 // daemonSet installs the calico/node container, as well as the Calico CNI plugins and network config on each
 // master and worker node in a Kubernetes cluster
-func daemonSet(ifacePatch bool, clusterCIDR, installCNIImage, calicoImage, flannelImage string) *appsv1.DaemonSet {
+func daemonSet(clusterCIDR, installCNIImage, calicoImage, flannelImage string) *appsv1.DaemonSet {
 	maxUnavailable := intstr.FromInt(1)
 	terminationGracePeriodSeconds := int64(0)
 	privileged := true
@@ -67,17 +67,14 @@ func daemonSet(ifacePatch bool, clusterCIDR, installCNIImage, calicoImage, flann
 				},
 			},
 		},
-	}
-
-	if ifacePatch {
-		flannelEnv = append(flannelEnv, corev1.EnvVar{
+		{
 			Name: "FLANNELD_IFACE",
 			ValueFrom: &corev1.EnvVarSource{
 				FieldRef: &corev1.ObjectFieldSelector{
 					FieldPath: "status.hostIP",
 				},
 			},
-		})
+		},
 	}
 
 	return &appsv1.DaemonSet{

--- a/pkg/templates/externalccm/ccm.go
+++ b/pkg/templates/externalccm/ccm.go
@@ -44,8 +44,6 @@ func Ensure(s *state.State) error {
 	s.Logger.Info("Ensure external CCM is up to date...")
 	var err error
 
-	s.PatchCNI = true
-
 	switch {
 	case s.Cluster.CloudProvider.Hetzner != nil:
 		err = ensureHetzner(s)

